### PR TITLE
PyTorch Geometric support in MeshGraphNet

### PR DESCRIPTION
**Pull request by Alexey-Kamenev**
# PhysicsNeMo Pull Request

## Description

The document describing the general approach is available [here](https://docs.google.com/document/d/1ze93EeYYAMEELIeNcFoZlE8PcTtAH8dsZnBKvpSnUzE).

## Key Changes

1. Added PyTorch Geometric (PyG) support in MeshGraphNet.
2. Added PyG-based vortex shedding example alongside the existing DGL implementation.
    * Added PyG-based `VortexSheddingDataset` datapipe.

## Features

* The changes are non-intrusive and do not affect the existing DGL implementation which users can continue to use.
* Existing DGL vortex shedding example has been renamed to `vortex_shedding_mgn_dgl`.
* A new vortex shedding example is added in `vortex_shedding_mgn`.
    * This approach allows users to easily compare different GNN implementations.
* Any existing commands or scripts that use the DGL implementation can be used with the new PyG implementation.
* Models are interchangeable between the DGL and PyG implementations.
* The changes are tested and validated. See Known Issues section for more details.

## Breaking Changes

* DGL datasets created previously as serialized DGLGraph objects will not be compatible 
  with the new PyG implementation. Users will need to change the corresponding datapipe
  to load DGLGraph objects and convert them to PyG objects.

## Out of scope

Advanced features are not included or not validated in this PR, such as:

* Distributed MGN.
* CuGraphOps

## Known Issues

* ~~DGL vs PyG loss curves in vortex shedding example. PyG loss curve has too much variance. [1379](https://github.com/NVIDIA/modulus-roadmap/issues/1379)~~

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

### CODE REVIEW

**Comment by coreyjadams** [comment]
nit: The requirements.txt should include torch_scatter.  I got it to work with 

```
pip install torch-scatter -f https://data.pyg.org/whl/torch-2.7.0+cu128.html
```

**Comment by Alexey-Kamenev** [comment]
> nit: The requirements.txt should include torch_scatter. I got it to work with

Good point - done!

File: examples/cfd/vortex_shedding_mgn_dgl/README.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
+# MeshGraphNet for transient vortex shedding
```
Comment (mnabian) [inline examples/cfd/vortex_shedding_mgn_dgl/README.md:1]: I suggest putting a note for deprecation.

File: physicsnemo/datapipes/gnn/vortex_shedding_dataset.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
 
 import numpy as np
 import torch
+import torch_geometric as pyg
 
 try:
     import tensorflow.compat.v1 as tf
```
Comment (mnabian) [inline physicsnemo/datapipes/gnn/vortex_shedding_dataset.py:27]: Once your PR is merged, I'll push a small PR to remove tensorflow dependency for data loading

File: physicsnemo/models/gnn_layers/utils.py [hunk 3]
```diff
@@ -327,9 +372,14 @@ def sum_efeat(
                 sum_efeat = update_efeat_bipartite_e2e(
                     efeat, src_feat, dst_feat, bipartite_graph, mode="sum"
                 )
-        else:
+        elif isinstance(graph, DGLGraph):
             src, dst = (item.long() for item in graph.edges())
             sum_efeat = sum_efeat_dgl(efeat, src_feat, dst_feat, src, dst)
+        elif isinstance(graph, PyGData):
+            src, dst = graph.edge_index.long()
+            sum_efeat = sum_efeat_dgl(efeat, src_feat, dst_feat, src, dst)
```
Comment (mnabian) [inline physicsnemo/models/gnn_layers/utils.py]: This was a bit confusing to me. We are using `sum_efeat_dgl` for PyG. I guess there is nothing dgl specific in that function, and I suggest renaming it.

File: examples/cfd/vortex_shedding_mgn_dgl/README.md (hunk 1 above)
Comment (Alexey-Kamenev) [reply examples/cfd/vortex_shedding_mgn_dgl/README.md:1]: Done.

File: physicsnemo/models/gnn_layers/utils.py (hunk 3 above)
Comment (Alexey-Kamenev) [reply physicsnemo/models/gnn_layers/utils.py]: Agree - I wanted to rename it but forgot to do that. Done.

File: examples/cfd/vortex_shedding_mgn/inference.py [hunk 4]
```diff
@@ -19,14 +19,14 @@
...
 
-from dgl.dataloading import GraphDataLoader
 import matplotlib.pyplot as plt
 from matplotlib import animation
 from matplotlib import tri as mtri
 from matplotlib.patches import Rectangle
 import numpy as np
 from omegaconf import DictConfig
 import torch
+from torch_geometric.loader import DataLoader
```
Comment (ktangsali) [inline examples/cfd/vortex_shedding_mgn/inference.py]: nit: from a user standpoint, this might create confusion with the `DataLoader` class from PyTorch. Wondering if we should call it `from torch_geometric.loader import DataLoader as PyGDataLoader` ?

File: examples/cfd/vortex_shedding_mgn/requirements.txt [hunk 5]
```diff
@@ -2,4 +2,6 @@ tensorflow<=2.17.1
 hydra-core>=1.2.0
 wandb>=0.13.7
 scipy>=1.15.0
-vtk>=9.2.6
\ No newline at end of file
+vtk>=9.2.6
+torch_geometric>=2.6.1
+torch_scatter>=2.1.2
```
Comment (ktangsali) [inline examples/cfd/vortex_shedding_mgn/requirements.txt:7]: Thanks for this! The PyG version can now remove the steps to install DGL: https://github.com/Alexey-Kamenev/modulus/tree/pyg-vortex-shedding/examples/cfd/vortex_shedding_mgn#prerequisites

File: physicsnemo/datapipes/gnn/vortex_shedding_dataset.py [hunk 6]
```diff
@@ -317,18 +301,19 @@ def create_graph(src, dst, dtype=torch.int32):
         creates a DGL graph from an adj matrix in COO format.
```
Comment (ktangsali) [inline physicsnemo/datapipes/gnn/vortex_shedding_dataset.py]: This comment needs an update

File: physicsnemo/models/gnn_layers/utils.py [hunk 7]
```diff
@@ -14,16 +14,21 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Any, Callable, Dict, Tuple, Union
+from typing import Any, Callable, Dict, Tuple, TypeAlias, Union
 
 import dgl.function as fn
 import torch
+import torch_scatter
 from dgl import DGLGraph
```
Comment (ktangsali) [inline physicsnemo/models/gnn_layers/utils.py]: I like the type alias solution! But just so I understand this correctly, presently to use the pyg only functionality, users still have to install DGL correct? Otherwise, this import will throw an error? In that case, you can ignore my previous comment about removing the DGL install from the pre-requisites section. However, it would be nice to make this explicit to the user. i.e., 

> post this change, to run the Vortex Shedding example (and maybe other GNN examples), **both** PyG and DGL are gonig to be required. 

Please correct me if my understanding is incorrect.

File: test/datapipes/test_vortex_shedding.py
Comment (ktangsali) [inline test/datapipes/test_vortex_shedding.py:1]: Thanks for adding this! I think this will help with the coverage 😉

**Review by ktangsali — APPROVED** [review]

File: examples/cfd/vortex_shedding_mgn/requirements.txt (hunk 5 above)
Comment (Alexey-Kamenev) [reply examples/cfd/vortex_shedding_mgn/requirements.txt:7]: Good catch - done! Technically it's still required during transition period but not by the example but the PhysicsNeMo MGN implementation.

File: examples/cfd/vortex_shedding_mgn/inference.py (hunk 4 above)
Comment (Alexey-Kamenev) [reply examples/cfd/vortex_shedding_mgn/inference.py]: Good point, done.

File: physicsnemo/datapipes/gnn/vortex_shedding_dataset.py (hunk 6 above)
Comment (Alexey-Kamenev) [reply physicsnemo/datapipes/gnn/vortex_shedding_dataset.py]: Done.

File: physicsnemo/models/gnn_layers/utils.py (hunk 7 above)
Comment (Alexey-Kamenev) [reply physicsnemo/models/gnn_layers/utils.py]: Yes, that's a very good point!
I've updated the code to make sure DGL sample does not require PyG and vice versa. Also added a bunch of warnings.

Comment (Alexey-Kamenev) [reply physicsnemo/models/gnn_layers/utils.py]: @ktangsali resolving for now, but please check later, specifically [this commit](https://github.com/NVIDIA/physicsnemo/pull/995/commits/2b424d77dcbdcb8559ea7d114325d2fa98526baa).
